core/unix/os.c

```diff
@@ -2012,7 +2012,7 @@ os_tls_thread_exit(local_state_t *local_state)
 
     /* We already set TLS to &uninit_tls in os_thread_exit() */
 
-    if (dynamo_exited && !last_thread_tls_exited) {
+    if (!client_thread && dynamo_exited && !last_thread_tls_exited) {
```

Comment:
Unless I understand the code path wrong, if a non-client-thread calls dr_exit_process, which calls to cleanup_and_terminate, and it will ends up the two synch-all, the flag will be set after the second synch-all since it is a non-client-thread.
The code here just try to avoid a client thread set the flag before the real last non-client-thread do the exit.